drivers/disk/flashdisk.c

```diff
@@ -284,16 +330,47 @@ static const struct disk_operations flash_disk_ops = {
 	.ioctl = disk_flash_access_ioctl,
 };
 
-static struct disk_info flash_disk = {
-	.name = CONFIG_DISK_FLASH_VOLUME_NAME,
-	.ops = &flash_disk_ops,
+#define DT_DRV_COMPAT fixed_partitions
+
+#define DEFINE_FLASHDISKS_MTD(node_id)						\
+{										\
+	.info = {								\
+		.name = DT_LABEL(node_id),					\
+		.ops = &flash_disk_ops,						\
+		.dev = DEVICE_DT_GET(DT_MTD_FROM_FIXED_PARTITION(node_id)),	\
+	},									\
+	.offset = DT_REG_ADDR(node_id),						\
+	.size = DT_REG_SIZE(node_id),						\
+},
+
+#define DEFINE_FLASHDISKS(inst)					\
+	DT_INST_FOREACH_CHILD(inst, DEFINE_FLASHDISKS_MTD)
+
+static struct mtd_disk_data mtd_disks[] = {
+	DT_INST_FOREACH_STATUS_OKAY(DEFINE_FLASHDISKS)
```

Comment:
Seems that you create list of all partitions from `fixed-partitions` compatible node and then search for the one that has the `"storage"` label to mark it as proper storage device.
I do not think that this is good idea, because there may be exactly one label `"storage"` in DTS which limits you to one storage type entry for flash, and it has to be named storage.
I am working, here https://github.com/zephyrproject-rtos/zephyr/pull/33144/commits, on macros to detect types for partitions and I would like to propose cooperation.
I would propose you to use compatible for marking partition as `storage` viable , for example:
```
compatible = "zephyr,storage-mtd";
```
or something like this. I will pursue to separate my commits that process such comatibles (`FLASH_AREA_FOR_EACH_COMPATIBLE`) so that they could be used for the purpose of your PR.
With use of such macros you would be able to list only partitions that would be marked as compatible with storage subsystem.

Reply:
Actually you do not even have to wait for my PR to go through; you need to propose new "compatible" string and change
```
#define DT_DRV_COMPAT fixed_partitions
```
to that string; then alter the  `DEFINE_FLASHDISKS_MTD`

Reply:
> Seems that you create list of all partitions from `fixed-partitions` compatible node and then search for the one that has the `"storage"` label to mark it as proper storage device.
> I do not think that this is good idea, because there may be exactly one label `"storage"` in DTS which limits you to one storage type entry for flash, and it has to be named storage.

All flash partitions should be found and mapped as block devices. What you mean by renaming "storage"->"NAND" is only necessary because ELM-FAT is so bad and depends on hard-coded names.

>I would propose you to use compatible for marking partition as storage viable , for example:
compatible = "zephyr,storage-mtd";
>Actually you do not even have to wait for my PR to go through; you need to propose new "compatible" string and change

That does not help here, it is not only about "storage" partition. As explained above all partitions should be mapped. "disk access" is the interface for subsystems that access block devices and is used by FS subsystem and USB MSC. Also USB DFU should use it in the future regardless on which block devices and how the updates are performed.

Reply:
Right, because the FAT driver actually decides by the mount point path where to go next, and the mount point list is limited to certain strings or single digits.
Yet, if you use "storage" partition to find flash device that will be used to fat, you limit the driver to single flash device, because you will have single "NAND" mount point that will be resolved, the first one on the flash list , right?
Because you may have only single label "storage" through the entire DTS configuration.
But, if you introduce compatible, you may have several compatibles, even on different chips, still there will be problem of "NAND" mount point, so the digits have to be used or the list within FAT driver configuration needs to be altered to take additional "NAND" points ("NAND0", "NAND1", etc.)
btw. NAND isn't even true, in most cases it is NOR.
And you do not have to register all flash partitions because you only need these you will be resolving with `disk_access_get_di` and in case of introducing the compatible you can strictly decide which one that will be.

Reply:
> Right, because the FAT driver actually decides by the mount point path where to go next, and the mount point list is limited to certain strings or single digits.
> Yet, if you use "storage" partition to find flash device that will be used to fat, you limit the driver to single flash device, because you will have single "NAND" mount point that will be resolved, the first one on the flash list , right?

Neither I nor flashdisk limits it. ELM-FAT and its interface to fs-subsytem is the limitation and this is not even fixable because of the usage of hard-coded names, that is just very bad. "storage" is the partition used for filesystem, this just follows the current state in Zephyr. Identifier for the block devices should be made differently and I am thinking about but problems with ELM-FAT remain anyway.

> Because you may have only single label "storage" through the entire DTS configuration.
> But, if you introduce compatible, you may have several compatibles, even on different chips, still there will be problem of "NAND" mount point, so the digits have to be used or the list within FAT driver configuration needs to be altered to take additional "NAND" points ("NAND0", "NAND1", etc.)
> btw. NAND isn't even true, in most cases it is NOR.

This is not a sustainable solution, it is just tinkering. "storage" partition is just one use case and has this special handling to provide workaround for ELM-FAT and may also be improved. Filesystem, whichever, should simply accept any valid block devices identifier.

> And you do not have to register all flash partitions because you only need these you will be resolving with `disk_access_get_di` and in case of introducing the compatible you can strictly decide which one that will be.

This is currently the case, all "fixed_partitions" compatible partitions are mapped.

Reply:
> Neither I nor flashdisk limits it. ELM-FAT and its interface to fs-subsytem is the limitation and this is not even fixable because of the usage of hard-coded names,

Ok, there has been probably miunderstanding of 

> Right, because the FAT driver actually decides by the mount point path where to go next, and the mount point list is limited to certain strings or single digits.

It is not an attack, it is statement of truth, the accent has been missing:  this is the culprit of our problem `_VOLUME_STRS`

Reply:
FWIW if the device volume was specified by `[0-9][0-9]*:` (FatFs numeric volume ID plus colon) using `FF_STR_VOLUME_ID=0` and Zephyr handled the mapping from mount point names to ids the problem would be solved.

Or the set of volume strings is configurable when FatFs is built, so providing a precise list is a possible solution, even if `get_ldnumber()` would have to be changed to reference an external volume list.

That Zephyr wants to use FatFs in a multi-filesystem environment but isn't adapting it to that use case isn't an entirely justifiable strike against FatFs.